pennylane/gradients/parameter_shift.py

```diff
@@ -558,115 +567,119 @@ def _process_pda2_involutory(tape, pdA2, var_idx, non_involutory):
+        pdA2 = _get_pdA2(
+            results[tape_boundary:], tape, pdA2_fn, non_involutory_indices, var_indices
+        )
 
+        # The logic follows:
+        # variances (var_mask==True): return d(var(A))/dp = d<A^2>/dp -2 * <A> * d<A>/dp
+        # plain expectations (var_mask==False): return d<A>/dp
+        # Note: if pdA2 != 0, then len(pdA2) == len(pdA)
         multi_measure = len(tape.measurements) > 1
+        num_params = len(tape.trainable_params)
         if multi_measure:
-            num_params = len(tape.trainable_params)
 
-            if num_params > 1:
+            if num_params == 1:
+
                 var_grad = []
+
                 for m_idx in range(len(tape.measurements)):
-                    m_res = []
-                    measurement_is_var = mask[m_idx]
-                    if measurement_is_var:
-                        for p_idx in range(num_params):
-                            _pdA2 = pdA2[m_idx][p_idx] if pdA2 != 0 else pdA2
-                            _f0 = f0[m_idx]
-                            _pdA = pdA[m_idx][p_idx]
-                            r = _get_var_with_second_order(_pdA2, _f0, _pdA)
-                            m_res.append(r)
-                        m_res = tuple(m_res)
-                    else:
-                        m_res = tuple(pdA[m_idx][p_idx] for p_idx in range(num_params))
+                    m_res = pdA[m_idx]
+                    if var_mask[m_idx]:
+                        _pdA2 = pdA2[m_idx] if pdA2 != 0 else pdA2
+                        _f0 = f0[m_idx]
+                        m_res = _get_var_with_second_order(_pdA2, _f0, m_res)
+
                     var_grad.append(m_res)
                 return tuple(var_grad)
 
-            p_idx = 0
             var_grad = []
             for m_idx in range(len(tape.measurements)):
-                measurement_is_var = mask[m_idx]
-                m_res = pdA[m_idx][p_idx]
-                if measurement_is_var:
-                    _pdA2 = pdA2[m_idx][p_idx] if pdA2 != 0 else pdA2
-                    _f0 = f0[m_idx]
-                    m_res = _get_var_with_second_order(_pdA2, _f0, m_res)
-
+                m_res = []
+                if var_mask[m_idx]:
+                    for p_idx in range(num_params):
+                        _pdA2 = pdA2[m_idx][p_idx] if pdA2 != 0 else pdA2
+                        _f0 = f0[m_idx]
+                        _pdA = pdA[m_idx][p_idx]
+                        r = _get_var_with_second_order(_pdA2, _f0, _pdA)
+                        m_res.append(r)
+                    m_res = tuple(m_res)
+                else:
+                    m_res = tuple(pdA[m_idx][p_idx] for p_idx in range(num_params))
                 var_grad.append(m_res)
             return tuple(var_grad)
 
-        # Scalar
-        var_grad = _get_var_with_second_order(pdA2, f0, pdA) if mask else pdA
-        return var_grad
+        # Single measurement case, meas has to be variance
+        if num_params == 1:

```

Comment:
Single measurement or single parameter?

Reply:
Single measurement case, then the `if` statement is specific to the single parameter case. The comment is meant to remind the reader that we've covered the multi-measurement case above.

Reply:
Thanks!